webpack.common.js

```diff
@@ -1,14 +1,72 @@
+const { resolve } = require('path');
 const webpack = require('webpack');
-const path = require('path');
 const { CleanWebpackPlugin } = require('clean-webpack-plugin');
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const CopyPlugin = require('copy-webpack-plugin');
+
+const pythonPath = resolve(__dirname, 'pydata_sphinx_theme');
+const staticPath = resolve(pythonPath, 'static');
+
+const vendor = resolve(staticPath, 'vendor');
+const vendorVersions = {
+  fontAwesome: require("@fortawesome/fontawesome-free/package.json").version,
+  openSans: require("@openfonts/open-sans_all/package.json").version,
+  lato: require("@openfonts/lato_latin-ext/package.json").version
+};
+const vendorPaths = {
+  fontAwesome: resolve(vendor, 'fontawesome', vendorVersions.fontAwesome),
+  openSans: resolve(vendor, 'open-sans_all', vendorVersions.openSans),
+  lato: resolve(vendor, 'lato_latin-ext', vendorVersions.lato)
+};
+
+// generates cache-busting templates to be used in `layout.html` without knowing versions
+function macroTemplate({compilation}){
+
+  const assets = Object.keys(compilation.assets);
+
+  const css = assets.filter((file) => file.endsWith(".css"));
+  const js = assets.filter((file) => file.endsWith(".js"));
+
+  const stylesheet = (css) => {
+    return `<link rel="stylesheet" href="{{ pathto('_static/${css}', 1) }}">`;
+  }
+
+  const preload = (js) => {
+    return `<link rel="preload" as="script" href="{{ pathto('_static/${js}', 1) }}">`
+  }
+
+  const script = (js) => {
+    return `<script src="{{ pathto('_static/${js}', 1) }}"></script>`;
+  }
+
+  return `<!-- these macros are generated by "yarn build". do not edit by hand. -->
+{% macro head_pre() %}
+  <link rel="stylesheet"
+    href="{{ pathto('_static/vendor/fontawesome/${vendorVersions.fontAwesome}/css/all.min.css', 1) }}">
+  <link rel="stylesheet"
+    href="{{ pathto('_static/vendor/open-sans_all/${vendorVersions.openSans}/index.css', 1) }}">
+  <link rel="stylesheet"
+    href="{{ pathto('_static/vendor/lato_latin-ext/${vendorVersions.lato}/index.css', 1) }}">
+  ${css.map(stylesheet).join('\n')}
+{% endmacro %}
+
+{% macro head_post() %}
+  ${js.map(preload).join('\n')}
+{% endmacro %}
+
+{% macro body_post() %}
+  ${js.map(script).join('\n')}
+{% endmacro %}`;
+}
+
 
 module.exports = {
   entry: {
     index: ['./src/js/index.js'],
   },
   output: {
-    filename: 'js/[name].js?[hash]',
-    path: path.resolve(__dirname, 'pydata_sphinx_theme/static'),
+    filename: 'js/[name].[hash].js',
```

Comment:
so that hash was getting dropped on the floor and not used anywhere. A production site without appropriate cache headers would happily keep serving `index.js` from the browser cache, even if the site had been rebuilt.

putting it in the filename is a sure-fire way to make it cause cache rebuilding: since it would _ship_ with that value, a downstream user would only incur that overhead when it gets reused.